src/typechecker/FStar.TypeChecker.TcTerm.fs

```diff
@@ -1205,91 +1212,121 @@ and tc_match (env : Env.env) (top : term) : term * lcomp * guard_t =
         | Some (e1, c1) -> e1, c1
         | None -> e1, c1 in
 
-    //ret_or_res_typ has type either<either<typ, comp>, typ>
-    let env_branches, ret_or_res_typ, ret_opt, g1 =
+    let env_branches, ret_opt, g1 =
       match ret_opt with
-      | None ->  //no return annotation, set an expected type either from env or a new uvar
+      | None ->
         (match Env.expected_typ env with
-         | Some t -> env, Inr t, None, g1
+         | Some _ -> env, None, g1
          | None ->
            let k, _ = U.type_u() in
            let res_t, _, g = TcUtil.new_implicit_var "match result" e1.pos env k in
-           Env.set_expected_typ env res_t, Inr res_t, None, Env.conj_guard g1 g)
-      | Some (t_or_c, None) ->
-        //typecheck the return annotation and unset expected type in the env if exists,
-        //  (we will check the expected type at the end)
-        //the branch typechecking (tc_eqn) will typecheck the branch with an ascription
+           Env.set_expected_typ env res_t,
+           None,
+           Env.conj_guard g1 g)
+      | Some (b, asc) ->
+        //We have a returns annotation,
+        //  clear the expected type in the environment for the branches
+        //  we will check the expected type for the whole match at the end
         let env, _ = Env.clear_expected_typ env in
-        let t_or_c, g =
-          match t_or_c with
-           | Inl t -> let k, _ = U.type_u () in
-                     let t, _, g = tc_check_tot_or_gtot_term env t k "" in
-                     Inl t, g
-           | Inr c -> let c, _, g = tc_comp env c in
-                     Inr c, g in
-        env, Inl t_or_c, Some (t_or_c, None), Env.conj_guard g1 g
-      | _ ->
-        raise_error (Errors.Fatal_UnexpectedTerm,
-          "Tactic is not yet supported with match return") (Env.get_range env) in
-
-    let guard_x, t_eqns =
-      let guard_x =
-        match ret_opt with
-        | None -> S.new_bv (Some e1.pos) c1.res_typ
-        | Some _ ->
-          (match (e1 |> U.unascribe |> SS.compress).n with
-           | Tm_name scrutinee_bv -> scrutinee_bv
-           | _ ->
-             raise_error (Errors.Fatal_UnexpectedTerm,
-               "The scrutinee must be a variable when a return annotation is supplied with a match") e1.pos) in
-      guard_x,
-      eqns |> List.map (tc_eqn guard_x env_branches ret_opt) in
+        let b, asc =
+          let bs, asc = SS.open_ascription [b] asc in
+          let b = List.hd bs in
+          //we set the sort of the binder to be the type of e1
+          {b with binder_bv={b.binder_bv with sort=c1.res_typ}}, asc in
```

Comment:
Adding the returns binder with sort as the type of `e1`, to typecheck the ascription.